compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -137,58 +173,190 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
     private fun coerceNotToUnit(fromType: Type, fromKotlinType: KotlinType?, toKotlinType: KotlinType): StackValue =
         codegen.coerceNotToUnit(fromType, fromKotlinType, toKotlinType)
 
-    private fun gen(
-        cases: ArrayList<Pair<Int, Label>>,
+    abstract inner class Switch(
+        val subject: IrGetValue,
+        val defaultLabel: Label,
+        val endLabel: Label,
+        val elseExpression: IrExpression?,
+        val thenExpressions: ArrayList<Pair<IrExpression, Label>>
+    ) {
+        open fun gen(): StackValue? {
+            genSubject()
+            genSwitch()
+            genThenExpressions()
+            val result = genElseExpression()
+
+            mv.mark(endLabel)
+            return result
+        }
+
+        protected abstract fun genSubject()
+
+        protected abstract fun genSwitch()
+
+        protected fun genIntSwitch(unsortedIntCases: List<Pair<Int, Label>>) {
+            val intCases = unsortedIntCases.sortedBy { it.first }
+            val caseMin = intCases.first().first
+            val caseMax = intCases.last().first
+            val rangeLength = caseMax - caseMin + 1L
```

Comment:
val rangeLength = caseMax.toLong() - caseMin.toLong() + 1L
It would be best to also add some test for overflow case